design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,123 @@
+* `SELECT` subqueries should not be allowed.
+
+#### Examples
+
+Given these restrictions, the following examples would be valid lambda functions:
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
+
+Lambda functions should require a specific **invocation function**, which tells ksqlDB how to apply the given lambda function to an input value, and what kind of result to return. The following invocation functions are proposed for the initial support for lambda functions:
+
+* `transform_array(arr, x => y)` - Applies the given lambda function to each element of the input `Array`, returning a new `Array` containing the transformed output.
+
+* `transform_keys(map, k => k)` - Applies the given lambda function to each key in the input `Map`, returning a new `Map` containing the transformed keys.
+
+* `transform_values(map, v => v)` - Applies the given lambda function to each value in the input `Map`, returning a new `Map` containing the transformed values.
+
+* `filter_array(arr, x => bool)` - Filters the input `Array` using the given lambda function. A new `Array` is returned, containing only values for which the lambda function evaluated to `true`.
+
+* `filter_map(map, (k, v) => bool)` - Filters the input `Map` using the given lambda function. A new `Map` is returned, containing only the key-value pairs for which the lambda function evaluated to `true`.
+
+* `reduce_array(arr, s, (x, s) => s)` - Reduces the input `Array` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_array` will return the final value of `s`.
+
+* `reduce_map(map, s, (k, v, s) => s)` - Reduces the input `Map` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_map` will return the final value of `s`.
+
+## Documentation updates
+
+We must document and provide examples for the two core aspects of lambda functions:
+
+1. Lambda function syntax
+2. Invocation functions
+
+## Compatibility implications
+
+Lambda functions would be additive to ksqlDB's grammar and should therefore not introduce any incompatibilities into the grammar or public APIs. However, it is not currently known if a lambda function implementation would require backward-incompatible changes to ksqlDB's internals.
+
+## Security implications
+
+Since lambda functions would only allow SQL expressions that can already be used within queries, no new security implications should be introduced.
+
+## Open questions
+
+~~**1. Would it be possible to allow aggregates in lambdas within the context of a particular invocation function? For example,**~~
```

Comment:
> COUNTIF(col, x -> shouldKeep)

Would lambdas enable any capabilities in this context? For example, what about this:
```
SELECT COUNT(CASE WHEN shouldKeep THEN col ELSE null END) ...
```
Having lambdas here would definitely simplify things but I'm trying to understand if they'd give users anything net new.

Reply:
I'd have to give it a think, I'm not sure I can come up with anything off the top of my head

Reply:
Would it not be possible to allow users to build their own UDAFs from lambdas by just providing a few build-in UDAFs called `udaf`:

```
-- Simple type:
-- col0: the input to aggregate
-- initial: initial agg value.
-- (col0, agg) => agg: the lambda that takes the current value and aggregate and calcs a new agg.
udaf(col0, initial, (col0, agg) => agg)

-- Extended to support different intermediate schema:
-- col0: the input to aggregate
-- initial: initial agg value.
-- (col0, m) => m: updates aggregate (with intermediate schema)
-- m => f, converts intermediate schema to final schema.
udaf(col0, initial, (col0, m) => m, m => f)

-- Session windows need a merge function too, so we'd need a session variant:
-- as above, plus:
-- (f1, f2) => merged: merge two sessions together
udaf(col0, initial, (col0, m) => m, (m1, m2) => merged, m => f)
```

I *think* just these few udafs would allow users to build custom aggregators using CASE statements for flow control, and build-in functions and operators.  

We can extend the above with multiple parameter variants:

```
udaf(col0, initial, (col0, agg) => agg)
udaf(col0, col1, initial, (col0, col1, agg) => agg)
udaf(col0, col2, col3, initial, (col0, col1, col3, agg) => agg)
... etc.
```

Of course, to do this we'd first need to do this: https://github.com/confluentinc/ksql/projects/4 (which I'm trying to get prioritized, as its a blocker for a _lot_ of UDAF improvements).